### ISSUE TITLE
[NPU] Add Qwen3.5-35B-A3B support

### ISSUE THREAD

**Issue by Fulin-Gao** [vllm-project/vime#389]
# Summary

This issue documents the end-to-end deployment and training workflow for running **Qwen3.5-35B-A3B** GRPO training with Vime on Ascend NPU (Atlas A3), intended as a reference for the community.

| Item | Version | Commit |
|---|---|---|
| Ascend HDK | `25.5.1.B030` | `N/A` |
| CANN | `9.0.0` | `N/A` |
| python | `3.12.13` | `N/A` |
| **torch_npu** | **`2.10.0.post2`** | **`N/A`** |
| triton-ascend | `3.2.1` | `N/A` |
| vllm | `0.22.1+empty` | `0decac0` |
| vllm-ascend | `0.22.1rc1` | `5f6faa0c` |
| **Megatron-Bridge** | **`N/A`** | **`7f0fb345`** |
| Megatron-LM | `0.16.0rc0` | `1dcf0dafa` |
| MegatronAdaptor | `0.1.0` | `15582ad` |
| TransformerEngineNPU | `2.13.0` | `d743c83` |
| **flash-linear-attention-npu** | **`v26.6.0`** | **`14c2c92`** |

# Usage
> ⚠️ Note: Before proceeding, you must update the following paths to match your server environment:
- Docker volume mappings in the container execution command
- Local storage directories designated for model and dataset downloads
- Internal paths and configurations defined within the training script

## 1. Setup

```bash
git clone -b ascend https://github.com/vllm-project/vime.git
cd vime
docker build -f docker/Dockerfile.npu -t vime-ascend:latest .
```

```bash
export IMAGE=vime-ascend:latest

docker run -d --name vime-npu -it --net=host --shm-size=1024g \
    --privileged=true \
    --cap-add=SYS_PTRACE \
    --device=/dev/davinci_manager \
    --device=/dev/hisi_hdc \
    --device=/dev/devmm_svm \
    -v /usr/local/Ascend/driver:/usr/local/Ascend/driver \
    -v /usr/local/dcmi:/usr/local/dcmi \
    -v /usr/local/bin/npu-smi:/usr/local/bin/npu-smi \
    -v /usr/local/sbin:/usr/local/sbin \
    -v /home:/home \
    -v /mnt:/mnt \
    -v /tmp:/tmp \
    -v /data:/data \
    -v /usr/share/zoneinfo/Asia/Shanghai:/etc/localtime \
    $IMAGE

docker exec -it vime-npu bash
```

## 2. Download

```bash
# (1) Model
hf download Qwen/Qwen3.5-35B-A3B --local-dir /path/to/Qwen3.5-35B-A3B

# (2) Data
hf download --repo-type dataset zhuzilin/dapo-math-17k --local-dir /path/to/dapo-math-17k
hf download --repo-type dataset zhuzilin/aime-2024  --local-dir /path/to/aime-2024
```

## 3. Training ([run-qwen3.5-35B-A3B-npu.sh](https://github.com/vllm-project/vime/blob/ascend/scripts/run-qwen3.5-35B-A3B-npu.sh))

```bash
cd /root/vime
bash scripts/run-qwen3.5-35B-A3B-npu.sh
```

# Results
<img width="642" height="348" alt="reward" src="https://github.com/user-attachments/assets/e06687bd-19c8-4c12-b9a7-ba8a9ed76772" />
<img width="642" height="348" alt="diff" src="https://github.com/user-attachments/assets/ffb0720b-df08-4692-b71a-a974ba8650ad" />
<img width="642" height="348" alt="eval" src="https://github.com/user-attachments/assets/50f3b8f3-5a53-4f4b-bfef-3b3d6672a48f" />

# Proposed Follow-ups
- [ ] **Integrate fla-npu native `causal_conv1d` operator**
Currently, the causal_conv1d in the Qwen3.5-35B-A3B GDN module uses the Triton implementation (megatron.core.ssm.triton.causal_conv1d).

- [ ] **Enable vLLM graph mode on Ascend (remove `--vllm-enforce-eager`)**
Currently, the rollout engine forces eager mode by hardcoding --vllm-enforce-eager in scripts/run-qwen3.5-35B-A3B-npu.sh due to unsupported NPU graph capture, disabling vLLM's graph-based decode acceleration.

- [ ] **Support dynamic batching for the GDN path (`--use-dynamic-batch-size`)**
Lacking Megatron GDN packing support, the training script is forced to use --micro-batch-size 1 with heavy padding, disabling dynamic batching.
